File: 112-path-sum/path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:


        # we are dfs we are goint perorder NLR
        def dfs(root,currSum):

            if not root: return False

            currSum+=root.val
            # check if node is  leadnode
            if not root.right and not root.left:
                # it is  should return true of not if return false
                return currSum == targetSum
            # left fails it gooes to right
            return dfs(root.right,currSum) or dfs(root.left,currSum)

        
        return dfs(root,0)

File: 112-path-sum/path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        std::function<bool(TreeNode*,int,int)> dfs = [&](TreeNode* root, int target ,int presentSum){
            if (!root){return false;}
            presentSum+=root->val;
            if(!root->left && !root->right){
                if(presentSum == target){return true; }
            }
            if( dfs(root->left,target,presentSum)){return true;}
            if(dfs(root->right,target,presentSum)){
            return true;
            }
            presentSum-=root->val;
            return false;
        };
      
        return dfs(root,targetSum,0);

    }
};